File: CPP/LRUCache.cpp
#include<stdio.h>
#include<bits/stdc++.h>
using namespace std;

class LRUCache{
    public:
        class node {
            public:
                int key;
                int val;
                node* next;
                node* prev;
                node(int _key, int _val){
                    key=_key;
                    val=_val;
                }
        };

        node* head= new node(-1, -1);
        node* tail = new node(-1, -1);

        int cap;

        unordered_map<int, node*> mp;

        LRUCache(int capacity){
            cap=capacity;
            head->next=tail;
            tail->prev=head;
        }

        void addnode(node* newnode){
            node* tmp = head->next;
            newnode->next=tmp;
            newnode->prev=head;
            head->next=newnode;
            tmp->prev=newnode;
        }

        void deletenode(node* delnode){
            node* deleteprev = deletenode->prev;
            node* deletenext = deletenode->next;
            deleteprev->next=deletenext;
            deletenext->prev=deleteprev;
            free(delnode);
        }

        int get(int key_){
            if(mp.find(key_)!=mp.end()){
                node* resnode=mp[key_];
                int res=resnode->val;
                mp.erase(key_);
                deletenode(resnode);
                addnode(resnode);
                mp[key_]=head->next;
                return res;
            }
        }

        void put(int key_, int value){
            if(mp.find(key_)!=mp.end()){
                node* existingnode = mp[key_];
                mp.erase(key_);
                deletenode(existingnode);
            }
            if(mp.size()==cap){
                mp.erase(tail->prev->key);
                deletenode(tail->prev);
            }
            addnode(new node(key_, value));
            mp[_key]=head->next;
        }
}

File: JAVA/LRUCache.java
package JAVA;
import java.util.*;
class LRUCache{
    Node head = new Node(0, 0), tail = new Node(0, 0);
    Map<Integer, Node> map=new HashMap<>();
    int capacity;

    public LRUCache(int _capacity){
        capacity=_capacity;
        head.next=tail;
        tail.prev=head;
    }

    public int get(int _key){
        if(map.containsKey(_key)){
            Node node = map.get(_key);
            remove(node);
            insert(node);
            return node.value;
        } else {
            return -1;
        }
    }

    public void put(int _key, int _value){
        if(map.containsKey(_key)){
            remove(map.get(_key));
        }
        if(map.size()==capacity){
            remove(tail.prev);
        }
        insert(new Node(_key, _value));
    }

    private void insert(Node node){
        map.put(node.key, node);
        Node headNext = head.next;
        head.next=node;
        node.prev=head;
        headNext.prev=node;
        node.next=headNext;
    }

    private void remove(Node node){
        map.remove(node.key);
        node.prev.next = node.next;
        node.next.prev=node.prev;
    }

    class Node{
        Node prev, next;
        int key, value;
        Node(int _key, int _value){
            key=_key;
            value=_value;
        }
    }
}